DOC-9214 udfs for query (#2513)
couchbase/docs-server: modules/guides/pages/javascript-udfs/creating-javascript-library.adoc

= Creating a JavaScript Library
:description: Creating a JavaScript library.
:page-pagination: 
:page-edition: Enterprise Edition
:page-topic-type: guide
:page-toclevels: 2

== Introduction

In this guide you will learn how to create a library for storing JavaScript functions, and how to add a new function to the library.

include::partial$javascript-udfs/what-you-need.adoc[tags="couchbase-setup"]

== Building the library

Javascript libraries can be created using the Query Workbench UI or the REST API.

> NOTE: You cannot create JavaScript libraries using the CBQ shell.

Reviewer: It's more accurate to say you cannot create JavaScript libraries using N1QL.

[tabs]
====
Query Workbench:: 
+
--
. Select btn:[Query] to access the Query Workbench, then select btn:[UDF] Query Workbench menu.
+
image::javascript-udfs/navigate-to-udf-query.png[alt="route to the user-defined functions screen"]

> . Click on the *{plus} add function library* link in the `Javascript Function Libraries` table to show the `Add Library` screen.

Reviewer: Bold for UI elements

```suggestion
. Click on the *{plus} add function library* link in the *Javascript Function Libraries* table to show the *Add Library* screen.
```

. Ensure the `Namespace` is set to `global`.

> . Enter a name for the library in the `Library Name` field.

Reviewer: Bold for UI elements. (Also reword this to incorporate scoped libraries.)

```suggestion
. Ensure the *Namespace* is set to `global`.
. Enter a name for the library in the *Library Name* field.
```

+
image::javascript-udfs/add-library-screen.png[,600, alt="add screen"]

. Add your own function to the library by typing it in below the preset example.
+

> image::javascript-udfs/added-get-business-days-function.png[,600, alt="added new function"]

Reviewer: We always use a code listing for the UI examples, rather than an image, so that users can copy and paste the example straight in. The example file, by necessity, does not need to be the same file as the one used for the REST API example.

. Save the library by pressing the btn:[Save] button.

> --

Reviewer: Include a link to your documentation for the UDF UI!

Author: I'm going to have to merge the files in from another branch for that.

REST API::
+
--
You will use the command shell `curl`command to execute REST commands.

Execute the following command to create a Javascript function.

[source, bash]
----
include::example$javascript-udfs/create-javascript-function-with-rest.sh[]
----
<.> In this example, the library is called `my-library`.
<.> Supply the username and password for your server.
<.> Supply the function for adding into the library.
--

> ====

Reviewer: Include a link to the Functions REST API documentation!

== Create the Reference Function

To call a JavaScript function, you first have to add a N1QL function to reference it.

> This can be done through the UI, the REST API or the CBQ Shell.

Reviewer: Again, more accurate to say:

```suggestion
This can be done through the UI, the REST API, or a N1QL query.
```

[tabs]
====
Query Workbench:: 
+
--

> . Click on the `{plus}{nbsp}add function` link from the `UDFs` screen.

Reviewer: Bold for UI elements

```suggestion
. Click on the *{plus}{nbsp}add function* link from the *UDFs* screen.
```

+
image::javascript-udfs/my-library-list-add-function-button.png[alt="library list"]
+

> which will display the `Add Function` screen.

Reviewer: ```suggestion
The *Add Function* dialog is displayed.
```

+
image::javascript-udfs/add-function-dialog.png[ ,400]

. Fill in the details as follows:
+
*Function Name*:: This is the name of the function that will be used in N1QL statements.

*Namespace*:: The function can be allocated a namespace that defines who can access it.
Users must have access rights to the namespace to run the function. 
The `global` namespace is set at the level of the bucket, so any N1QL statement run by a user with access to the bucket will be able to run the function. 
We can set the namespace to a particular scope, ensuring that the function is only available within that scope.

*Parameters*:: It is possible to explicitly define the list of parameters the function is expecting. 
However, the default setting  `...` indicates a variable length parameter list, which will be matched to the JavaScript parameter list when  the function is executed.

*Function Type*:: Set this to `Javascript`. 

> *Javascript Library*:: This of name of the JavaScript library where your function is stored.

Reviewer: ```suggestion
*Javascript Library*:: The name of the JavaScript library where your function is stored.
```

> *Library Function Name*:: The name of the JavaScript function in `my-library` that this definition will call.

Reviewer: ```suggestion
*Library Function Name*:: The name of the JavaScript function in the specified library that this definition will call.
```

. Press the btn:[Save Function] button to save the definition.

NOTE: In the background, The dialog will use the gathered information to generate and execute a N1QL xref:n1ql:n1ql-language-reference/createfunction.adoc[`CREATE FUNCTION`] call to generate the function definition.

--

REST API::
+
--
Add the reference function to your bucket using a REST call executed from the command shell:

[source, console]
----
include::example$javascript-udfs/create-reference-function.sh[]
----
[sidebar]
=====
include::partial$javascript-udfs/create-function-breakdown.adoc[]
=====
--

> CBQ Shell::

Reviewer: ```suggestion
{sqlpp}::
```

+
--
include::partial$javascript-udfs/start-cbq-shell-instructions.adoc[]
. Enter the N1QL statement to create the reference function:
+

> [source, console]

Reviewer: ```suggestion
[source, n1ql]
```
